### ISSUE TITLE
restricting plots to x and/or y ranges

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#1195]
We occasionally get users who want to know how to restrict a plot to a certain x or y range. This can be done "after the fact" - e.g. with something like

    for ax in plt.gcf().axes:
        ax.set_xlim(2, 12)

(or just, for a single plot, `plt.gca().set_ylim(2e-4, 3e-2)`, but I think it's worthwhile enough to be able to do with our plot infrastructure. Indeed, we already support this for the UI `plot_source` routine - but only for the X axis - with its `lo` and `hi` arguments.

Things we could do:

- decide whether it's something we want to support or just require "post processing" by the user
- if we do take it up, do we want to add `lo`/`hi` arguments to all the `plot_xxx`/`get_xxx_plot` routines (and probably `ylo`/`yhi`) or do we want to make it a preference setting like `color`, `xlog`, ... [which for the user doesn't really make a difference; it may not even make much of a difference to the code]
- we could argue about whether to have `lo` and `hi` separate or require a two-element sequence to set both at once, but I think we just stick with the former.

Straw-person idea:

```python
>>> from sherpa.astro.ui import *
...
>>> plot_fit(ylog=True, ylo=1e-4)
>>> plot_model(lo=0.1, hi=10)

# not sure about this
>>> plot_data(xlog=True, ylog=True, lo=0.1, hi=10)
>>> plot_model(overplot=True)
>>> plot_model_component(pl, lo=2, hi=8, overplot=True)
```

The main implementation issue is is this just like the proposed work-around - create a plot of all the data and then limit it after drawing it - or do we only want to evaluate the data on the requested range [to some degree of fidelity given binning/grouping/filtering] and display that - which is a bit of what the last example is wondering about.

**Comment by hamogu** [comment]
I vote for post-processing (and using that in a few more of the Sherpa threads as example). My reasoning is mostly to prevent feature-creep and save us the maintenance and the users the complexity of the expanded list of parameters; I think this is more complex than it looks at at first. If we add `x_lo, x_hi`, then we probably also should add `y_lo, y_hi`. But then what do we do for plots that have two axes (e.g. fit and del_chi)? Applying the same `y_lo, y_hi` to both of them makes no sense. So, should that be tuples or lists like `(y_lo_1, y_hi_1), (y_lo_2, y_hi_2)` or maybe `(y_lo_1, y_lo_2), (y_hi_1, y_hi_2)`? Or do we always upply it to the data plot and don't have an option for the chi^2-plot?

However, this is probably the most commonly used option for extra formatting, so I could be convinced to add it to our plotting directly for easy cases.

A related thought: When I plot e.g. fit and chi^2, I get two axes objects. Are the x-axes linked for those? I can't really see a case where someone would want to zoom into the spectrum in x-direction, but not into the Chi^2. If they are linked, then an easy way to get the effect the user wants is to just manually zoom with the mouse in an interactive matplotlib window (either the gui or "matplotlib notebook"  in Jupyter and then press the "save" button to export the plot. If the user wants to make zoomed plots in a script or program, they should be comfortable calling matplotlib commands.

**Comment by DougBurke** [comment]
For multi-plot cases like `plot_fit_delchi` I don't believe they are linked (note that this becomes a backend-specific issue since our hypothetical other-backend plot case may not have a "shared axis" case). For future reference I think it's the https://sherpa.readthedocs.io/en/latest/plots/api/sherpa.plot.JointPlot.html class we'd care about, but it may be one of those cases where the split between generic and backend-specific functionality may make this hard to achieve.

**Comment by hamogu** [comment]
There already is a line `backend.set_jointplot(0, 0, self.rows, self.cols, create=create)` in there. I think that's where a backend (if possible) joins the axes. And, for the hypothetical second backend that I'm working on (on and off) that's very much possible: https://docs.bokeh.org/en/latest/docs/user_guide/interaction/linking.html#userguide-interaction-linking
(It's also possible in the other possible backends I've considered. That's not to say that every library has it, but it's so common that essentially every mature library has it.)

**Comment by DougBurke** [comment]
Getting off-topic a not here, but I know you've been looking at the backend stuff - should we add a section of the developer notes to ensure we have a `dummy` module (name to be bike-shedded) for anything that has a backend in which we set up the interface we require a backend to provide? We have this for plotting, although it's woefully undocumented for this use case, and I think you just added an io version in #1185, or have I mis-understood that PR?

**Comment by hamogu** [comment]
That is correct. We always had a dummy backend for datastack (which had it's own IO backend) ,s this really is nothing new. I will end documentation in some form (either in the dummy backend or in a separate page when I'm adding a new plotting backend. For the IO, I'm not sure I'm in deep enough to write those docs yet, so I would rather do that as part of #1091 than #1185, because #1091 will require me to understand how to add routines that actually do something.

**Comment by hamogu** [comment]
Related: #316
